operators/src/pro/cache/shared_cache.rs

```diff
@@ -0,0 +1,1190 @@
+    // TODO: use a multi-dimensional index to speed up the lookup
+    entries: HashMap<CacheEntryId, C>,
+
+    // running queries insert their tiles as they are produced. The entry will be created once the query is done.
+    // The query is identified by a Uuid instead of the query rectangle to avoid confusions with other queries
+    landing_zone: HashMap<QueryId, L>,
+}
+
+impl<C, L> OperatorCacheEntry<C, L> {
+    pub fn new() -> Self {
+        Self {
+            entries: Default::default(),
+            landing_zone: Default::default(),
+        }
+    }
+
+    fn insert_landing_zone_entry(
+        &mut self,
+        query_id: QueryId,
+        landing_zone_entry: L,
+    ) -> Result<(), CacheError> {
+        let old_entry = self.landing_zone.insert(query_id, landing_zone_entry);
+
+        if old_entry.is_some() {
+            Err(CacheError::QueryIdAlreadyInLandingZone)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn remove_landing_zone_entry(&mut self, query_id: &QueryId) -> Option<L> {
+        self.landing_zone.remove(query_id)
+    }
+
+    fn landing_zone_entry_mut(&mut self, query_id: &QueryId) -> Option<&mut L> {
+        self.landing_zone.get_mut(query_id)
+    }
+
+    fn insert_cache_entry(
+        &mut self,
+        cache_entry_id: CacheEntryId,
+        cache_entry: C,
+    ) -> Result<(), CacheError> {
+        let old_entry = self.entries.insert(cache_entry_id, cache_entry);
+
+        if old_entry.is_some() {
+            Err(CacheError::CacheEntryIdAlreadyInCache)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn remove_cache_entry(&mut self, cache_entry_id: &CacheEntryId) -> Option<C> {
+        self.entries.remove(cache_entry_id)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.entries.is_empty() && self.landing_zone.is_empty()
+    }
+
+    fn iter(&self) -> impl Iterator<Item = (&CacheEntryId, &C)> {
+        self.entries.iter()
+    }
+}
+
+identifier!(QueryId);
+
+impl ByteSize for QueryId {}
+
+identifier!(CacheEntryId);
+
+impl ByteSize for CacheEntryId {}
+
+/// Holds all the tiles for a given query and is able to answer queries that are fully contained
+#[derive(Debug, Hash)]
+pub struct CacheQueryEntry<Query, Elements> {
+    query: Query,
+    elements: Elements,
+}
+type RasterOperatorCacheEntry = OperatorCacheEntry<RasterCacheQueryEntry, RasterLandingQueryEntry>;
+type RasterCacheQueryEntry = CacheQueryEntry<RasterQueryRectangle, CachedTiles>;
+type RasterLandingQueryEntry = CacheQueryEntry<RasterQueryRectangle, LandingZoneQueryTiles>;
+
+type VectorOperatorCacheEntry = OperatorCacheEntry<VectorCacheQueryEntry, VectorLandingQueryEntry>;
+type VectorCacheQueryEntry = CacheQueryEntry<VectorQueryRectangle, CachedFeatures>;
+type VectorLandingQueryEntry = CacheQueryEntry<VectorQueryRectangle, LandingZoneQueryFeatures>;
+
+impl<Query, Elements> CacheQueryEntry<Query, Elements> {
```

Comment:
if you can implement everything exactly the same (for each generic impl), do we even need this trait?